enhancements/sdk-external-and-pluggable-validations.md

```diff
@@ -0,0 +1,494 @@
+
+TODO: need to define what the CLI for operator-sdk will look like. Are there
+any new flags? Do we use the environment variable?
+
+NOTE: what would the CLI look like?
+
+
+##### Default validators run by operator-sdk
+
+List of current validators:
+TODO: do we need these validators called out like this?
+
+* BundleValidator
+  * validates the bundle
+  * looks for duplicate keys in bundle
+  * verifies all owned keys match a CRD in the bundle
+  * verifies that all CRDs present in the bundle are in the CSV
+  * validates the service account
+
+* ClusterServiceVersionValidator
+  * operator-sdk checks to see if the CSV is nil on the bundle (that's the first
+    check)
+  * checks that the CSV name is a valid format
+    * DNS1123
+    * valid label
+  * verifies replaces name is also a valid format
+    * DNS1123
+    * valid label
+  * ensures that both `alm-examples` and `olm.examples` are not both present
+  * decodes the example yaml, to validate its format
+  * checks provided APIs
+  * validates the `InstallModes`
+    * verifies that conversion CRDs support `AllNamespaces`
+  * checks for missing mandatory fields
+  * validates the annotation names
+  * validates the version and kind
+
+* CustomResourceDefinitionValidator
+  * operator-sdk puts the v1beta1 and v1 CRDs together for validation
+  * validates v1beta1 CRDs
+  * validates v1 CRDs
+  * validates internal CRDs
+    * https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/validation/validation.go#L49-L78
+
+##### Optional validators
+
+* PackageManifestValidator
+
+* OperatorHubValidator
+
+* ObjectValidator
+
+* OperatorGroupValidator
+
+* CommunityOperatorValidator
+
+* AlphaDeprecatedAPIsValidator
+
+* AllValidators
+  * Uses ALL of the validators
+    * BundleValidator
+    * ClusterServiceVersionValidator
+    * CustomResourceDefinitionValidator
+    * PackageManifestValidator
+    * OperatorHubValidator
+    * ObjectValidator
+    * OperatorGroupValidator
+    * CommunityOperatorValidator
+    * AlphaDeprecatedAPIsValidator
+
+* DefaultBundleValidators
+  * Uses the following 3 validators
+    * BundleValidator
+    * ClusterServiceVersionValidator
+    * CustomResourceDefinitionValidator
+### Risks and Mitigations
+
+There is little risk, if this does not pan out we keep going on the current path
+of compiling them in.
+
+## Design Details
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
```

Comment:
IHMO the external validators ought to be responsible to test their rules. 
SDK just need to ensure and test that still able to work well with the external validators feature.

Reply:
Yes, removed the above generic template wording and replaced it with this:
```
* unit testing of the sdk external validator feature
* bundle validate e2e
* external validators will have their own test coverages
```